### DISCUSSION TITLE
Is it possible to get multi-node rootless k3s working?

### DISCUSSION THREAD

**Discussion by harsimranmaan** [k3s-io/k3s#9869] [General]
I installed rootless k3s server and want to join a rootless agent from another node. Are there any tips on how to go about it?Agent comes up but cannot bootstrap properly. I see that the docs say that multi-node is not supported. I am curious if someone can point me in the right direction on where to look and guide if it is possible at all to achieve multinode given that usernetes supports it.
 I used the following systemd config. 
```
cat > .config/systemd/user/k3s-rootless.service <<EOF
[Unit]
Description=k3s  agent (Rootless)

[Service]
Environment=PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
ExecStart=/usr/local/bin/k3s agent --data-dir /home/ubuntu/.rancher/k3s --server https://172.31.33.102:6443 --token xyz --rootless --snapshotter=fuse-overlayfs
ExecReload=/bin/kill -s HUP \$MAINPID
TimeoutSec=0
RestartSec=2
Restart=always
StartLimitBurst=3
StartLimitInterval=60s
LimitNOFILE=infinity
LimitNPROC=infinity
LimitCORE=infinity
TasksMax=infinity
Delegate=yes
Type=simple
KillMode=mixed

[Install]
WantedBy=default.target
EOF
```

Error
```
pr 04 22:15:20 ip-172-31-40-190 k3s[1632]: time="2024-04-04T22:15:20Z" level=info msg="containerd is now running"
Apr 04 22:15:20 ip-172-31-40-190 k3s[1632]: time="2024-04-04T22:15:20Z" level=info msg="Getting list of apiserver endpoints from server"
Apr 04 22:15:20 ip-172-31-40-190 k3s[1632]: time="2024-04-04T22:15:20Z" level=info msg="Updated load balancer k3s-agent-load-balancer default server address -> 10.41.0.100:6443"
Apr 04 22:15:20 ip-172-31-40-190 k3s[1632]: time="2024-04-04T22:15:20Z" level=info msg="cgroup v2 controllers are delegated for rootless."
Apr 04 22:15:20 ip-172-31-40-190 k3s[1632]: time="2024-04-04T22:15:20Z" level=info msg="Running kubelet --address=0.0.0.0 --allowed-unsafe-sysctls=net.ipv4.ip_forward,net.ipv6.conf.all.forwarding --anonymous-auth=false --authentication-token-webhook=true --authorization-mode=Webhook --cert-dir=/home/ubuntu/.rancher/k3s/agent/kubelet/pki --cgroup-driver=cgroupfs --client-ca-file=/home/ubuntu/.rancher/k3s/agent/client-ca.crt --cloud-provider=external --cluster-dns=10.43.0.10 --cluster-domain=cluster.local --container-runtime-endpoint=unix:///run/k3s/containerd/containerd.sock --containerd=/run/k3s/containerd/containerd.sock --eviction-hard=imagefs.available<5%,nodefs.available<5% --eviction-minimum-reclaim=imagefs.available=10%,nodefs.available=10% --fail-swap-on=false --feature-gates=CloudDualStackNodeIPs=true --feature-gates=KubeletInUserNamespace=true --healthz-bind-address=127.0.0.1 --hostname-override=ip-172-31-40-190 --kubeconfig=/home/ubuntu/.rancher/k3s/agent/kubelet.kubeconfig --kubelet-cgroups=/k3s --node-ip=10.41.0.100 --node-labels= --pod-infra-container-image=rancher/mirrored-pause:3.6 --pod-manifest-path=/home/ubuntu/.rancher/k3s/agent/pod-manifests --read-only-port=0 --resolv-conf=/etc/resolv.conf --root-dir=/home/ubuntu/.rancher/k3s/agent/kubelet --runtime-cgroups=/k3s --serialize-image-pulls=false --tls-cert-file=/home/ubuntu/.rancher/k3s/agent/serving-kubelet.crt --tls-private-key-file=/home/ubuntu/.rancher/k3s/agent/serving-kubelet.key"
Apr 04 22:15:20 ip-172-31-40-190 k3s[1632]: time="2024-04-04T22:15:20Z" level=info msg="Connecting to proxy" url="wss://10.41.0.100:6443/v1-k3s/connect"
Apr 04 22:15:20 ip-172-31-40-190 k3s[1632]: time="2024-04-04T22:15:20Z" level=error msg="Failed to connect to proxy. Empty dialer response" error="dial tcp 10.41.0.100:6443: connect: connection refused"
Apr 04 22:15:20 ip-172-31-40-190 k3s[1632]: time="2024-04-04T22:15:20Z" level=error msg="Remotedialer proxy error" error="dial tcp 10.41.0.100:6443: connect: connection refused"
Apr 04 22:15:20 ip-172-31-40-190 k3s[1632]: time="2024-04-04T22:15:20Z" level=info msg="Waiting to retrieve kube-proxy configuration; server is not ready: failed to get CA certs: Get \"https://127.0.0.1:6444/cacerts\": read tcp 127.0.0.1:34250->127.0.0.1:6444: read: connection reset by peer"
Apr 04 22:15:25 ip-172-31-40-190 k3s[1632]: time="2024-04-04T22:15:25Z" level=info msg="Connecting to proxy" url="wss://10.41.0.100:6443/v1-k3s/connect"
Apr 04 22:15:25 ip-172-31-40-190 k3s[1632]: time="2024-04-04T22:15:25Z" level=error msg="Failed to connect to proxy. Empty dialer response" error="dial tcp 10.41.0.100:6443: connect: connection refused"
Apr 04 22:15:25 ip-172-31-40-190 k3s[1632]: time="2024-04-04T22:15:25Z" level=error msg="Remotedialer proxy error" error="dial tcp 10.41.0.100:6443: connect: connection refused"
Apr 04 22:15:25 ip-172-31-40-190 k3s[1632]: time="2024-04-04T22:15:25Z" level=info msg="Waiting to retrieve kube-proxy configuration; server is not ready: failed to get CA certs: Get \"https://127.0.0.1:6444/c
```

**Comment by brandond** [comment]
That would require additional development. PRs would be welcome.

From the docs:
> Multi-node rootless clusters, or multiple rootless k3s processes on the same node, are not currently supported. See [#6488](https://github.com/k3s-io/k3s/issues/6488#issuecomment-1314998091) for more details.

There is discussion in that issue as to what needs work.

**Reply by harsimranmaan** [reply]
Thanks @brandond. Are you referring to https://github.com/k3s-io/k3s/issues/6488#issuecomment-1315011810
```
Flannel (VXLAN) is known to work with rootless, but not integrated to rootless k3s yet.
```
Any pointers on where to look for this in the code?

**Reply by brandond** [reply]
I would probably talk to the usernetes maintainers about that. Most of the current rootless support was contributed by @AkihiroSuda . That is his comment you're asking about.

**Reply by AkihiroSuda** [reply]
Here: https://github.com/rootless-containers/usernetes/tree/gen2-v20240404.1
- Unlike k3s and Usernetes "Gen1", the current Usernetes Gen2 is wrapped inside Rootless Docker/Podman/nerdctl for simplicity.
  Maybe k3s should follow the same approach, possibly using k3d.
- The VXLAN port 8472/udp is exposed here: https://github.com/rootless-containers/usernetes/blob/gen2-v20240404.1/docker-compose.yaml#L21
- The node IP is not directly accessible from other nodes.
  "External IP" is set here: https://github.com/rootless-containers/usernetes/blob/gen2-v20240404.1/Makefile.d/sync-external-ip.sh#L12
- Flannel has an issue about UDP checksums for VXLAN behind NAT.
  A workaround is applied: https://github.com/rootless-containers/usernetes/blob/gen2-v20240404.1/Dockerfile.d/etc_udev_rules.d_90-flannel.rules
- The vanilla `kube-flannel.yaml` can be used: https://github.com/rootless-containers/usernetes/blob/gen2-v20240404.1/Makefile#L137

**Comment by brandond** [comment]
> Usernetes Gen2 is wrapped inside Rootless Docker/Podman/nerdctl for simplicity.
Maybe k3s should follow the same approach, possibly using k3d.

I think we would prefer that it remain usable standalone. If it becomes only available when running inside a container, that would  probably be even less usable than the current state.

**Reply by AkihiroSuda** [reply]
Flannel works with Usernetes "Gen1" (standalone) too, like this way
https://github.com/rootless-containers/usernetes/blob/gen1/boot/flanneld.sh
